dwave/optimization/src/nodes/numbers.cpp

```diff
@@ -28,110 +33,311 @@ std::span<const Update> NumberNode::diff(const State& state) const noexcept {
+    }
+    return output_vec;
 }
 
-// Binary Node
+BinaryNode::BinaryNode(std::span<const ssize_t> shape,
+                       std::optional<std::vector<double>> lower_bound,
+                       std::optional<std::vector<double>> upper_bound)
+        : IntegerNode(shape, limit_bound_to_bool_domain<false>(lower_bound),
+                      limit_bound_to_bool_domain<true>(upper_bound)) {}
+
+BinaryNode::BinaryNode(std::initializer_list<ssize_t> shape,
+                       std::optional<std::vector<double>> lower_bound,
+                       std::optional<std::vector<double>> upper_bound)
+        : BinaryNode(std::span(shape), lower_bound, upper_bound) {}
+BinaryNode::BinaryNode(ssize_t size, std::optional<std::vector<double>> lower_bound,
+                       std::optional<std::vector<double>> upper_bound)
+        : BinaryNode({size}, lower_bound, upper_bound) {}
+
+BinaryNode::BinaryNode(std::span<const ssize_t> shape, double lower_bound,
+                       std::optional<std::vector<double>> upper_bound)
+        : BinaryNode(shape, std::vector<double>{lower_bound}, upper_bound) {}
+BinaryNode::BinaryNode(std::initializer_list<ssize_t> shape, double lower_bound,
+                       std::optional<std::vector<double>> upper_bound)
+        : BinaryNode(std::span(shape), std::vector<double>{lower_bound}, upper_bound) {}
+BinaryNode::BinaryNode(ssize_t size, double lower_bound,
+                       std::optional<std::vector<double>> upper_bound)
+        : BinaryNode({size}, std::vector<double>{lower_bound}, upper_bound) {}
+
+BinaryNode::BinaryNode(std::span<const ssize_t> shape,
+                       std::optional<std::vector<double>> lower_bound, double upper_bound)
+        : BinaryNode(shape, lower_bound, std::vector<double>{upper_bound}) {}
+BinaryNode::BinaryNode(std::initializer_list<ssize_t> shape,
+                       std::optional<std::vector<double>> lower_bound, double upper_bound)
+        : BinaryNode(std::span(shape), lower_bound, std::vector<double>{upper_bound}) {}
+BinaryNode::BinaryNode(ssize_t size, std::optional<std::vector<double>> lower_bound,
+                       double upper_bound)
+        : BinaryNode({size}, lower_bound, std::vector<double>{upper_bound}) {}
 
-void BinaryNode::flip(State& state, ssize_t i) const {
+BinaryNode::BinaryNode(std::span<const ssize_t> shape, double lower_bound, double upper_bound)
+        : BinaryNode(shape, std::vector<double>{lower_bound}, std::vector<double>{upper_bound}) {}
+BinaryNode::BinaryNode(std::initializer_list<ssize_t> shape, double lower_bound, double upper_bound)
+        : BinaryNode(std::span(shape), std::vector<double>{lower_bound},
+                     std::vector<double>{upper_bound}) {}
+BinaryNode::BinaryNode(ssize_t size, double lower_bound, double upper_bound)
+        : BinaryNode({size}, std::vector<double>{lower_bound}, std::vector<double>{upper_bound}) {}
+
+bool BinaryNode::flip(State& state, ssize_t i) const {
```

Comment:
I think you could write this as
```c++
if (lower_bound(i) == upper_bound(i)) {
    // variable is fixed
    return false;
}

ptr->set(i, !ptr->get(i));
return true;
```

Reply:
This assumes the bounds are only ever 0.0 or 1.0, which I believe will be the case?